include/psa/crypto.h

```diff
@@ -4043,6 +4043,553 @@ psa_status_t psa_generate_random(uint8_t *output,
+ *                              must be initialized first, and have had \c
+ *                              psa_sign_hash_start() called with it first.
+ *
+ * \param[out] signature        Buffer where the signature is to be written.
+ * \param signature_size        Size of the \p signature buffer in bytes. This
+ *                              must be appropriate for the selected
+ *                              algorithm and key:
+ *                              - The required signature size is
+ *                                #PSA_SIGN_OUTPUT_SIZE(\c key_type, \c
+ *                                key_bits, \c alg) where \c key_type and \c
+ *                                key_bits are the type and bit-size
+ *                                respectively of key.
+ *                              - #PSA_SIGNATURE_MAX_SIZE evaluates to the
+ *                                maximum signature size of any supported
+ *                                signature algorithm.
+ * \param[out] signature_length On success, the number of bytes that make up
+ *                              the returned signature value.
+ *
+ * \retval #PSA_SUCCESS
+ *         Operation completed successfully
+ *
+ * \retval #PSA_OPERATION_IN_PROGRESS
+ *         Operation was interrupted due to the setting of \c
+ *         psa_interruptible_set_max_ops(), there is still work to be done,
+ *         please call this function again with the same parameters.
+ *
+ * \retval #PSA_ERROR_BUFFER_TOO_SMALL
+ *         The size of the \p signature buffer is too small. You can
+ *         determine a sufficient buffer size by calling
+ *         #PSA_SIGN_OUTPUT_SIZE(\c key_type, \c key_bits, \p alg)
+ *         where \c key_type and \c key_bits are the type and bit-size
+ *         respectively of \p key.
+ *
+ * \retval #PSA_ERROR_NOT_SUPPORTED
+ * \retval #PSA_ERROR_INVALID_ARGUMENT
+ * \retval #PSA_ERROR_INSUFFICIENT_MEMORY
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
+ * \retval #PSA_ERROR_STORAGE_FAILURE
+ * \retval #PSA_ERROR_DATA_CORRUPT
+ * \retval #PSA_ERROR_DATA_INVALID
+ * \retval #PSA_ERROR_INSUFFICIENT_ENTROPY
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_sign_hash_complete( psa_sign_hash_operation_t *operation,
+                                     uint8_t *signature, size_t signature_size,
+                                     size_t *signature_length );
+
+/** Abort a sign hash operation.
+ *
+ * Aborting an operation frees all associated resources except for the
+ * \p operation structure itself. Once aborted, the operation object
+ * can be reused for another operation by calling
+ * \c psa_sign_hash_operation_init() again.
```

Comment:
start, not init

Reply:
The `init` takes 'raw' memory and turns into a usable operation object object. Prior to calling `init`, using any function that expects the memory in the object to be meaningful could have unspecified consequences.

Once initialized, any other function should leave the operation object in a valid state for the object - which might be 'operation reset'.

`abort` should leave the object in a state that is 'similar to' the state after `init`, and ready to call `start`. There is no need to call `init` after the first use.

Reply:
This is my assumption about the operation object memory state model:

```mermaid
stateDiagram-v2 TD
[*] --> IDLE: <code>init()
note left of [*]
   We start with uninitialised
   memory where the operation
   object will be
end note
state begin <<choice>>
state progress <<choice>>
IDLE --> begin: <code>start()
begin --> ACTIVE: <em>SUCCESS
begin --> IDLE: <em>ERROR
ACTIVE --> progress: <code>complete()
progress --> ACTIVE: <em>INCOMPLETE
progress --> IDLE: <em>SUCCESS
progress --> IDLE: <em>ERROR
```

IDLE and ACTIVE obviously have more complex internal state that track the progress of the operation, and the observable `num_ops()` attribute, but these are not shown in this high-level model.

Reply:
The best analogy is probably thinking of `init()` as an object constructor (pick your favourite language that has these) - you only run the constructor once to initialise the memory for the object. You can then use the object any number of times.

Reply:
Many thanks for this explanation, its extremely useful. I have now fixed this.